drivers/video/gc2145.c

```diff
@@ -0,0 +1,1148 @@
+	}
+
+	k_msleep(30);
+
+	return 0;
+}
+
+static int gc2145_set_resolution(const struct device *dev, uint16_t img_width, uint16_t img_height)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+
+	uint16_t win_w = UXGA_HSIZE + 16;
+	uint16_t win_h = UXGA_VSIZE + 8;
+	uint16_t c_ratio = 1;
+	uint16_t r_ratio = 1;
+	uint16_t x = (((win_w / c_ratio) - UXGA_HSIZE) / 2);
+	uint16_t y = (((win_h / r_ratio) - UXGA_VSIZE) / 2);
+	uint16_t win_x = ((UXGA_HSIZE - win_w) / 2);
+	uint16_t win_y = ((UXGA_VSIZE - win_h) / 2);
+
+	/* Set readout window first. */
+	ret = gc2145_set_window(dev, GC2145_REG_BLANK_WINDOW_BASE, win_x, win_y, win_w, win_h);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set cropping window next. */
+	ret = gc2145_set_window(dev, GC2145_REG_WINDOW_BASE, x, y, 1, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable crop */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_CROP_ENABLE, GC2145_CROP_SET_ENABLE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set Sub-sampling ratio and mode */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_SUBSAMPLE, ((r_ratio << 4) | c_ratio));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_SUBSAMPLE_MODE, GC2145_SUBSAMPLE_MODE_SMOOTH);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_msleep(30);
+
+	return 0;
+}
+
+static uint8_t gc2145_check_connection(const struct device *dev)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+	uint8_t reg_pid_val;
+	uint8_t reg_ver_val;
+
+	ret = gc2145_read_reg(&cfg->i2c, 0xf0, &reg_pid_val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_read_reg(&cfg->i2c, 0xf1, &reg_ver_val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((reg_ver_val != GC2145_REV_VAL) || (reg_pid_val != GC2145_PID_VAL)) {
+		LOG_ERR("failed to detect GC2145 pid: 0x%x rev: 0x%x", reg_pid_val, reg_ver_val);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int gc2145_set_fmt(const struct device *dev, enum video_endpoint_id ep,
+			  struct video_format *fmt)
+{
+	struct gc2145_data *drv_data = dev->data;
+	uint16_t width, height;
+	int ret;
+	int i = 0;
+
+	/* We only support RGB565 formats */
+	if (fmt->pixelformat != VIDEO_PIX_FMT_RGB565) {
+		LOG_ERR("gc2145 camera supports only RGB565");
+		return -ENOTSUP;
+	}
+
+	width = fmt->width;
+	height = fmt->height;
+
+	if (!memcmp(&drv_data->fmt, fmt, sizeof(drv_data->fmt))) {
+		/* nothing to do */
+		return 0;
+	}
+
+	drv_data->fmt = *fmt;
+
+	/* Set output format */
+	ret = gc2145_set_output_format(dev, fmt->pixelformat);
+	if (ret < 0) {
+		LOG_ERR("Failed to set the output format");
+		return ret;
+	}
+
+	/* Check if camera is capable of handling given format */
+	while (fmts[i].pixelformat) {
+		if (fmts[i].width_min == width && fmts[i].height_min == height &&
+		    fmts[i].pixelformat == fmt->pixelformat) {
+			/* Set window size */
+			ret = gc2145_set_resolution(dev, fmt->width, fmt->height);
+			if (ret < 0) {
+				LOG_ERR("Failed to set the resolution");
+			}
+			return ret;
+		}
+		i++;
+	}
```

Comment:
It looks like it is safe to convert this to a `for (int i = 0; ...; i++)` as `i` is used nowhere else.
Maybe some refactor on the original driver this got inspired from.